src/models/thread.ts

```diff
@@ -317,31 +322,33 @@ export class Thread extends ReadReceipt<EmittedEvents, EventHandlerMap> {
             this.replyCount = bundledRelationship.count;
             this._currentUserParticipated = !!bundledRelationship.current_user_participated;
 
-            const mapper = this.client.getEventMapper();
             // re-insert roomId
-            this.lastEvent = mapper({
+            this.lastEvent = new MatrixEvent({
                 ...bundledRelationship.latest_event,
                 room_id: this.roomId,
             });
+            this.updatePendingReplyCount();
             await this.processEvent(this.lastEvent);
         }
+    }
 
-        let pendingEvents: MatrixEvent[];
-        if (this.pendingEventOrdering === PendingEventOrdering.Detached) {
-            pendingEvents = this.room
-                .getPendingEvents()
-                .filter((ev) => ev.isRelation(THREAD_RELATION_TYPE.name) && this.id === ev.threadRootId);
-            await Promise.all(pendingEvents.map((ev) => this.processEvent(ev)));
-        } else {
-            pendingEvents = this.events
-                .filter((ev) => ev.isRelation(THREAD_RELATION_TYPE.name))
-                .filter((ev) => ev.status !== EventStatus.SENT && ev.status !== EventStatus.CANCELLED);
-        }
+    private updatePendingReplyCount(): void {
+        const unfilteredPendingEvents =
+            this.pendingEventOrdering === PendingEventOrdering.Detached ? this.room.getPendingEvents() : this.events;
+        const pendingEvents = unfilteredPendingEvents.filter(
+            (ev) =>
+                ev.threadRootId === this.id &&
+                ev.isRelation(THREAD_RELATION_TYPE.name) &&
+                ev.status !== null &&
+                ev.getId() !== this.lastEvent?.getId(),
```

Comment:
This looks awfully similar to what is in `onEcho`, any chance to consolidate the two under one roof?

Reply:
They do look similar, but so far they're not similar enough to easily unify them :/